[docs] Write a `decrease test execution time` article (#5895)
DevExpress/testcafe: docs/articles/documentation/recipes/best-practices/create-faster-tests.md

---
layout: docs
title: Create Faster Tests
permalink: /documentation/recipes/best-practices/create-faster-tests.html
redirect_from:
  - /documentation/recipes/create-faster-tests.html
---
# Speed Up Your Tests

> The speed of TestCafe tests depends on the loading speed of your web application. The best way to speed up your tests may be to optimize the tested application. Make sure that your application doesn't load large payloads of unused CSS or scripts.

Reviewer: Place ways in such order (from most valuable):

- use the `concurrency` option. Decrease time due to parallelization. It requires more resources because few browser instances will be used.
- use headless browsers. Decrease the test page loading time due to avoiding of the display the rendered page.
- use 'Roles 'feature
- use the fastest speed
- run in local browsers
- mock requests
- optimize your test code (page model, remove unnecessary modules, etc.)

'Optimize your application' is a bad suggestion for the 'Create faster steps' article.

Test in an environment with performance headroom. Lack of resources may increase the time browsers take to initialize, connect to TestCafe and load the tested application. If you have to test in a resource-low environment, [run tests in headless browsers](#run-tests-in-headless-browsers).

> For better performance, launch tests in local browsers. Network lag between TestCafe and remote browsers causes delays in test execution.

Author: Nice one, I like it

> Remote browsers may take a long time to initialize and connect to TestCafe. Use the  [--browser-init-timeout](../../reference/command-line-interface.md#--browser-init-timeout-ms) CLI option to set a time limit for browsers to connect. Tests fail if this timeout is exceeded - it might be a good time to debug your remote browser connection.

Author: This is a questionable point. I am unsure if it is viable.

@wentwrong  what do you think?

Reviewer 2: I think it's better to remove this part at all, because it doesn't seem to be related to the disclosed topic of the article.

Author: Gone

> TestCafe's [speed](../../reference/command-line-interface.md#--speed-factor) option allows you to change test speed. The default value is `1`, which is fastest.

Reviewer 2: I don't have any insight into how the "speed" option actually implemented, but from my point of view the "speed" option specified in the CLI, API, configuration file or set using the [setTestSpeed](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/settestspeed.html) method overrides the corresponding option of the emulation action (the speed with which this action occur) and affects the speed of the emulated cursor movement (the delay between the emulation actions, in fact).

I would approach this part in a different way:
> TestCafe can emulate a "human-like" interaction with a tested page by imitating delays with which certain actions are emulated on the page. It can be achieved by specifying the "speed" option on the test run level (CLI, API, config file), the test level (t.setTestSpeed) or level of the separate action. But if your concern is speed of test execution, you probably don't want waste time on it. So you should check that the "speed" option is set to the maximum possible value - `1`, which is set by default.

> TestCafe's [speed](../../reference/command-line-interface.md#--speed-factor) option allows you to change test speed. The default value is `1`, which is fastest.

Author: > TestCafe emulates real user actions on tested webpages. [Speed](../../reference/command-line-interface.md#--speed-factor) option enables you to change action emulation speed.

> The highest value of `1` represents the fastest actions on the page. This is a default value. The minimum value of `0.01` represents a speed that is 100 lower. Setting a lower `speed` can be useful for debugging tests, but slows tests down.

> If your tests have `speed` set in the run configuration, disable this setting or set it to `1`.

If your tests have `speed` set in the run configuration, disable this setting or set it to `1`.

## Run Tests in Headless Browsers

> Run tests in headless browsers. Headless browsers take less time to initialize because they don't need to render the application to a GUI. They also enable you to run tests in environments that lack GUI capabilities, like CI containers.

Author: ```suggestion
Run tests in headless browsers. Headless browsers take less time to initialize because they don't need to render the application. They also enable you to run tests in environments without graphical capabilities, such as CI containers.
```

To run tests in a headless browser, use the `:headless` CLI parameter:

```sh
testcafe "chrome:headless" tests/

> ```

Author: I guess we do provide a link to the corresponding sections in the docs.

For more information, see [Test in Headless Mode](../../guides/concepts/browsers.md#test-in-headless-mode)

## Use Roles for Login

If your tests require activity from logged in users, use [Roles](../../guides/advanced-guides/authentication.md#user-roles) for authentication.

> Put Roles in the [beforeEach hook](../../reference/test-api/fixture/beforeeach.md) of your test suite: following tests in a suite reuse authentication data and authentication happens instantly.

Author: Right

```suggestion
Roles are user log-in routines. When you first enable a `Role`, TestCafe executes the routine and saves the user authentication data that it receives. When you activate the same role later, TestCafe loads the authentication data and skips the log-in process. 
```

> With `Roles`, TestCafe remembers cookies associated with every logged-in account. When you switch Roles, TestCafe erases authentication data and you don't have to log out manually.

Author: Auth data might be in the `localStorage` as well

```suggestion
When you switch between `Roles`, TestCafe clears your browser's existing authentication data and loads the new role's credentials.
```

If your test covers logins on multiple websites, the active Role collects the authentication data from all these websites. When you switch to that Role later, you are logged in to all these websites.

Use an [Anonymous Role](../../guides/advanced-guides/authentication.md#anonymous-role) to instantly log out of all accounts during a test.

> > Roles can access authentication data in cookie and browser storage. If your authentication system stores data elsewhere, you may not be able to use roles.

Author: Weird phrasing. Perhaps:

```suggestion
> Roles can access authentication data in cookie and browser storage. If your authentication system stores data elsewhere, roles may not work.
```

This is [from our docs ](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/authentication.html#why-use-roles:~:text=can%20access%20authentication%20data%20in%20cookie%20and%20browser%20storage.%20If%20your)btw. Might as well change it there too

For more info on Roles, see [User Roles](../../guides/advanced-guides/authentication.md#user-roles)

## Run Tests Concurrently

Tests execute faster if run concurrently. In concurrent mode, TestCafe creates multiple instances of specified browsers and uses that browser pool to run tests.

Use the [--concurency](../../reference/command-line-interface.md#-c-n---concurrency-n) CLI option to launch tests concurrently:

```sh
testcafe --concurency 3 chrome tests/
```

For more info on concurrency, read [Run Tests Concurrently](../../guides/basic-guides/run-tests.md#run-tests-concurrently).

## Mock Requests

> A tested application may interact with remote resources (for example, an analytics service or a database). Requests to such resources may create delays in test suite execution. To avoid such delays, you can mock requests to these resources.

Reviewer 3: The first sentence is obvious and unnecessary 
```suggestion
You can mock the responses of time-consuming HTTP requests to speed up the testing process. 
```

> TestCafe [request mocker](../../reference/test-api/requestmock/README.md) intercepts requests from your app to external resources and responds with data that you specify. A mocked request is resolved almost instantly, which eliminates a possible delay caused by a foreign resource or network lag.

Reviewer 3: minor stylistic edits.
foreign is a bad word in this context, too imprecise. (also see: alien). an external resource is not necessarily foreign (for example if we're talking about a different server on the same CDN)
```suggestion
TestCafe [request mocker](../../reference/test-api/requestmock/README.md) intercepts your application's requests to external resources, and responds with the data that you specify. A mocked request is resolved almost instantly, which eliminates a possible delay caused by data processing and network lag.
```

Author: I like `latency` better

```suggestion
TestCafe [request mocker](../../reference/test-api/requestmock/README.md) intercepts your application's requests to external resources, and responds with the data that you specify. A mocked request is resolved almost instantly, which eliminates a possible delay caused by data processing and network latency.
```

For more information about mocking requests, visit [Mock HTTP Requests](../../guides/advanced-guides/intercept-http-requests.md#mock-http-requests).

## Optimize Your Page Model

> Page model is a great way to structure your test suite. However, wrong page model structure can increase testing time in large test suites.

Reviewer 3: A test suite is a set of testing tools. Why can't we use the word 'fixture' ?
*Wrong* implies something that does not reflect reality. 'Improper' fits this particular sentence much better.

```suggestion
Page models offer a great way to outline your application. However, improper page model structure can increase testing time in large fixtures.
```

Author: [Test suite is a collection of tests.](https://softwaretestingfundamentals.com/test-suite/#post-2042:~:text=A%20TEST%20SUITE%20is%20a%20collection%20of%20test%20cases.)

> In your page model file, do not export the page model constructor, but create and export a new instance.

Reviewer 3: Logic
```suggestion
Do not export the page model constructor itself. Use the constructor to create a new page model instance, and export that instance.
```

Author: It's actually a class, right?

```suggestion
Do not export the page model class. Create and export a new page model instance.
```

```js
import { Selector } from 'testcafe';
import { t } from 'testcafe';

class MyPage {

>     //page model contents here

Reviewer 3: ```suggestion
    //page model contents go here
```

Author: ```suggestion
    //page model contents
```

}

export default new MyPage()
// creates an instance of MyPage
// that can be imported in tests
```

> In your test files, avoid the `new` keyword when creating page model instances. Import the page object instead:

Reviewer 3: ```suggestion
Avoid the `new` keyword when creating page model instances. Import the page object instead:
```

```js
import myPage from 'path/to/page-model.js'

> // a recommended way to import page models

Reviewer 3: ```suggestion
// the recommended way to import page models
```

Author: [Same](https://github.com/DevExpress/testcafe/pull/5895#discussion_r565552816-permalink:~:text=important%20to%20mention%20that%20we're%20now%20in%20a%20page%20model%20file), but for a test file

//
const myPage = new MyPage();
// NOT RECOMMENDED

> // may lead to performance issues in large test suites

Reviewer 3: ```suggestion
// may lead to performance issues in large fixtures
```

Author: A test suite can (and often does) contain multiple fixtures. This approach would only significantly benefit a test suite of [with numerous page models and dozens of tests or more](https://github.com/DevExpress/testcafe/issues/4054#issuecomment-519450486)

---
layout: docs
title: Create Faster Tests
permalink: /documentation/recipes/best-practices/create-faster-tests.html
redirect_from:
  - /documentation/recipes/create-faster-tests.html
---
# Speed Up Your Tests

This section describes ways to decrease the execution time of TestCafe tests.

> 

Reviewer: What about adding the list of all described ways to provide customer a brief overview:

```js
* [run tests concurrently](http://link1)
* [run tests in headless browsers](http://link2)

```

Author: I thought the article would be too small for that kind of thing. But it has grown quite significantly, so it's a good suggestion
